packages/amplify-category-function/src/provider-utils/awscloudformation/utils/lambda-layer-cloudformation-template.ts

```diff
@@ -76,20 +64,100 @@ export default function generateLayerCfnObj(parameters) {
           Ref: 'AWS::Region',
         },
       },
-      Permission: {
-        Value: {
-          Ref: 'LambdaLayerPermission',
-        },
-      },
     },
   };
+  let cfnObj = { ...generateLayerCfnObjBase(), ...outputObj };
+  const POLICY_RETAIN = DeletionPolicy.Retain;
+  const layer = new Lambda.LayerVersion({
+    CompatibleRuntimes: parameters.runtimes.map(runtime => runtime.cloudTemplateValue),
+    Content: {
+      S3Bucket: Fn.Ref('deploymentBucketName'),
+      S3Key: Fn.Ref('s3Key'),
+    },
+    Description: 'Lambda Layer',
+    LayerName: joinWithEnv('-', parameters.layerName),
+    LicenseInfo: 'MIT',
+  });
+  layer.deletionPolicy(POLICY_RETAIN);
+  _.assign(layer, { UpdateReplacePolicy: POLICY_RETAIN });
 
-  if (layerVersionPermission) {
-    layerVersionPermission.deletionPolicy(POLICY_RETAIN);
-    layerVersionPermission.UpdateReplacePolicy = POLICY_RETAIN;
-    _.assign(layerVersionPermission, { UpdateReplacePolicy: POLICY_RETAIN });
+  cfnObj.Resources['LambdaLayer'] = layer;
 
-    cfnObj.Resources = { ...cfnObj.Resources, ...layerVersionPermission };
-  }
+  // parameters.laatestLayerVersion is defined
+  const layerData = layerMetadataFactory(context, parameters.layerName);
+  Object.entries(parameters.layerVersionsMap).forEach(([key]) => {
+    const answer = assignLayerPermissions(layerData, key, parameters.layerName, parameters.build);
+    answer.forEach(permission => (cfnObj.Resources[permission.name] = permission.policy));
+  });
   return cfnObj;
 }
+
+function assignLayerPermissions(layerData: LayerMetadata, version: string, layerName: string, flag: boolean) {
+  const layerVersionPermissionBase = {
+    Action: 'lambda:GetLayerVersion',
+    LayerVersionArn: createLayerversionArn(layerData, layerName, version, flag),
+  };
+
+  const result = [];
+  if (layerData.getVersion(Number(version)).isPublic()) {
+    result.push({
```

Comment:
I would early return in the isPublic and isPrivate condition because they are mutually exclusive